### ISSUE TITLE
Clean: improve performance

### ISSUE THREAD

**Issue by brandonlockhart** [sfu-db/dataprep#441]
## Summary

Make the dataprep.clean APIs performant.

## Design-level Explanation Actions

- [x] Can we assume the data fits in memory? **No.**
- [ ] What performance are we looking to achieve?

## Implementation-level Explanation

If we want to parallelize the computations, I think we need to use a low-level language, or Dask and be mindful of the GIL. Two approaches:

1. **Parallelized Cython:** Process the data in a for loop perhaps using [prange](https://cython.readthedocs.io/en/latest/src/userguide/parallelism.html) for parallelization. The data would need to be accessible by the native code without holding the GIL; we can consider using [memoryviews](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html) of a numpy array, eg the conversion pd.DataFrame -> np.ndarray -> memoryview (Cython), but I'm not sure if this is a good [approach for strings](https://stackoverflow.com/questions/28774096/cython-memory-view-of-ndarray-of-strings-or-direct-ndarray-indexing). 
Also, using [C strings in Cython could be inefficient](https://cython.readthedocs.io/en/latest/src/tutorial/strings.html#general-notes-about-c-strings) in our setting; we can instead use C++ wstrings.
    * **Pros**:
        * Could be very fast
    * **Cons**:
        * The string processing may be complex, thus more prone to bugs and more difficult to maintain

2. **Dask Processes:** The data could be cleaned using separate processes, which can [avoid GIL issues](https://docs.dask.org/en/latest/setup/single-machine.html#single-machine-scheduler). This could be better if we want to clean the data in partitions.
    * **Pros**:
        * Can still use Python to process the strings
    * **Cons**:
        * The dask overhead could make processing smaller dataframes slow 
        * When the kernel is interrupted in jupyter, the main process stops but the workers may continue to run. Thus for the next run of the cell, it might be stuck waiting for the lock.

Other things to consider:
a. Caching. In `clean_country`, the result of each cleaned country is cached; we need to be careful about doing this in parallel. We might need a separate cache for each thread. 

## Rationale and Alternatives

* **Numba** is suggested for improving performance, however, it may [not be good](https://pandas.pydata.org/pandas-docs/stable/user_guide/enhancingperf.html#caveats) for string operations. 

## Additional Tasks

- [x] This task is put into a correct pipeline (Development Backlog or In Progress).
- [x] The label of this task is setting correctly.
- [x] The issue is assigned to the correct person.
- [x] The issue is linked to related Epic.

- [ ] The documentation is changed accordingly.
- [ ] Tests are added accordingly.

A large speed increase can be achieved by iterating over a series rather than using dataframe.apply. Other changes in the result below include using a Pandas dataframe, and not using a global dictionary for the stats.

![dataprep_clean_lat_long_speedup](https://user-images.githubusercontent.com/25874021/101546562-d4364400-395d-11eb-84e2-252e692bfa52.png)

@dovahcrow contributed a lot to the above ideas

**Comment by brandonlockhart** [comment]
Below are some more experiments. Young mentioned that the `apply` function has some overhead, so I also tried a list comprehension: `[transform(value) for value in df["column"]` (for dask, I used the list comp on each partition using `.map_partitions()`), which could be [more efficient than apply](https://stackoverflow.com/a/55557758). However, I found for both pandas and dask that **`apply` and list comps perform similarly**. The results below used list comps. 

I used the function `clean_lat_long`, and transformed the degrees-minutes-seconds format (e.g., `"33° 56′ 20.796″ N, 67° 42′ 35.8308″ E"`) into two columns (latitude and longitude) each containing degrees as floats. I used a local dask cluster with 4 workers and 2 threads per worker and ran the experiments on the HPC. I used 128MB partitions like we use in EDA.

![Screen Shot 2020-12-20 at 1 14 52 AM](https://user-images.githubusercontent.com/25874021/102709780-3e09f400-4262-11eb-9a84-0ec4caef4917.png)

I'm unsure about why using separate processes performs significantly worse on small data (I switch from threads to processes using the [`scheduler` parameter](https://docs.dask.org/en/latest/setup/single-machine.html#selecting-threads-processes-or-single-threaded) in [dask.compute](https://docs.dask.org/en/latest/api.html#dask.compute)). Also, since the pandas dataframe fits comfortably in memory (there shouldn't be any swapping), dask seems to be parallelizing the computations using threads (the dashboard below is from the experiment with 10 million rows) which is unexpected since from my understanding the [GIL is held](https://docs.dask.org/en/latest/setup/single-machine.html#single-machine-scheduler) since we're performing string operations. I can investigate this further, or any ideas @dovahcrow or others?

![Screen Shot 2020-12-20 at 12 41 16 AM](https://user-images.githubusercontent.com/25874021/102710413-ef128d80-4266-11eb-9c17-221a878ba6df.png)


**Next steps:** Unless anyone sees any problems or has any other ideas for experiments, I suggest for now that we switch to dask with threads. This will require minor changes (~20 lines per file) to the current code. 

### Out-of-memory data

To avoid loading the entire dataframe into memory, I think it's possible to iteratively process the partitions and write them back to the disk (see [point 1 here](https://stackoverflow.com/a/61975950)). I think we could add some additional parameters to each `clean_{type}()` function (denoting whether to load the cleaned data into memory or not). I'm unsure about how often people are using larger-than-memory data, so I think this is a lower priority task, but I'm open to other opinions.

**Comment by brandonlockhart** [comment]
Addressed in #466, #506, #507
